cs/src/core/io/safe/OutputBuffer.cs

```diff
@@ -233,7 +236,20 @@ internal virtual void Grow(int count)
                 length = minLength;
             }
 
-            Array.Resize(ref buffer, length);
+            buffer = ResizeBuffer(buffer, length);
+            length = buffer.Length;
+        }
+
+        /// <summary>
+        /// Resize the internal buffer.
+        /// </summary>
+        /// <param name="buffer">Existing buffer.</param>
+        /// <param name="newSize">New buffer size.</param>
+        /// <returns>The new buffer.</returns>
+        protected virtual byte[] ResizeBuffer(byte[] buffer, int newSize)
```

Comment:
Thanks for considering it. You're right: let's skip adding an interface the allocation routines interface.

Thinking about the future, collecting some different implementations of `IOutputStream` and `IInputStream` to the Bond.IO.CSharp package could be a good addition. E.g., a segmented buffer implementation of `IOutputStream`.

You're right. We don't want to add `IDisposable` to either `OutputBuffer` or `InputStream`: it would be a breaking behavioral change, and the lifetime management would be super complex, especially for `InputStream`.

The `ActiveAllocationChunk` code is there as a denial of service defense to limit how much memory is allocated when reading a blob. Before this code was added, a payload that said it had a 1 GiB blob would cause a 1 GiB allocation. Now, it will only allocate up to `ActiveAllocationChunk` at a time, and to cause 1 GiB of allocation you have to have 1 GiB of payload. There is similar code to control [pre-allocation of container elements](https://github.com/microsoft/bond/blob/2fab2130615b91d317ac28464a25f9dd7d74c5d0/cs/src/core/expressions/DeserializerTransform.cs#L8).

The `ArrayPool` interface would work here, with the caveat the Bond cannot guarantee that it will return everything rented. Bond doesn't know when the buffers are no longer used.

In theory, adding a dependency on the System.Buffers package for &lt; netstandard2.1 would be fine. However, there are enough consumers of Bond that don't use NuGet and the additional dependency would be a breaking change.